propsigner/propsigner.go

```diff
@@ -0,0 +1,142 @@
+package propsigner
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/filecoin-project/go-state-types/crypto"
+	"github.com/filecoin-project/specs-actors/actors/builtin/market"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	pb "github.com/textileio/go-auctions-client/gen/wallet"
+	logger "github.com/textileio/go-log/v2"
+)
+
+const (
+	v1Protocol            = "/auctions/proposal-signer/1.0.0"
+	maxRequestMessageSize = 100 << 10 // 100KiB
+
+	// filecoinDealProtocolV1 refers exactly to the libp2p protocol used to send deal proposals.
+	// This value is used to know how to unmarshal proposal payloads. Future deal proposal versions
+	// might be supported, and thus we need to be able to distinguish them to do proper unmarshaling.
+	filecoinDealProtocolV1 = "/fil/storage/mk/1.1.0"
+)
+
+var (
+	log            = logger.Logger("propsigner")
+	streamDeadline = time.Minute
+
+	errInvalidAuthToken  = errors.New("invalid auth token")
+	errWalletMissingKeys = errors.New("wallet doesn't have keys for address")
+)
+
+// Wallet contains private keys for Filecoin addresses.
+type Wallet interface {
+	Has(addr string) (bool, error)
+	Sign(addr string, payload []byte) (*crypto.Signature, error)
+}
+
+type dealSignerService struct {
+	authToken string
+	wallet    Wallet
+}
+
+// NewDealSignerService configures a stream handler for the proposal signer protocol.
+func NewDealSignerService(h host.Host, authToken string, wallet Wallet) error {
+	if authToken == "" {
+		return fmt.Errorf("authorization token is empty")
+	}
+	dss := dealSignerService{
+		authToken: authToken,
+		wallet:    wallet,
+	}
+	h.SetStreamHandler(v1Protocol, dss.streamHandler)
+
+	return nil
+}
+
+func (dss *dealSignerService) streamHandler(s network.Stream) {
+	log.Infof("executing stream handler")
+	defer func() {
+		if err := s.Close(); err != nil {
+			log.Errorf("closing deal proposal signer stream: %s", err)
+		}
+	}()
+	if err := s.SetDeadline(time.Now().Add(streamDeadline)); err != nil {
+		log.Errorf("set deadline in stream: %s", err)
+	}
+
+	var req pb.ProposalSigningRequest
+	if err := readMsg(s, maxRequestMessageSize, &req); err != nil {
+		replyWithError(s, "unmarshaling proposal signing request: %s", err)
+		return
+	}
+	if req.AuthToken != dss.authToken {
+		replyWithError(s, errInvalidAuthToken.Error())
+		return
+	}
+
+	var walletAddr string
+	switch req.FilecoinDealProtocol {
+	case filecoinDealProtocolV1:
```

Comment:
Here's the future-proof support of newer versions of deal proposals. Today we only have one, but it would be easy to add newer ones if appear.